src/stores/disk.rs

```diff
@@ -0,0 +1,712 @@
+        self
+    }
+
+    /// Return the disk path, or load from the env var: CACHED_DISK_PATH, or fall back to OS cache directory
+    pub fn disk_path(&self) -> Result<PathBuf, DiskCacheBuildError> {
+        match self.disk_path {
+            Some(ref s) => Ok(s.to_path_buf()),
+            None => match std::env::var(ENV_KEY) {
+                Ok(path) => Ok(PathBuf::from(path)),
+                Err(error) => {
+                    let disk_path = BaseDirs::new().map(|base_dirs|
+                        base_dirs
+                            .cache_dir()
+                            .join(env!("CARGO_PKG_NAME"))
+                            .join("cached")
+                    );
+
+                    match disk_path {
+                        Some(path) => Ok(path),
+                        None => Err(DiskCacheBuildError::MissingDiskPath {
+                            env_key: ENV_KEY.to_string(),
+                            error,
+                        })
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn build(self) -> Result<DiskCache<K, V>, DiskCacheBuildError> {
+        let disk_path = self.disk_path()?;
+        let connection = sled::open(disk_path.clone())?;
+
+        Ok(DiskCache {
+            connection,
+            seconds: self.seconds,
+            refresh: self.refresh,
+            disk_path,
+            namespace: self.namespace,
+            prefix: self.prefix,
+            _phantom_k: self._phantom_k,
+            _phantom_v: self._phantom_v,
+        })
+    }
+}
+
+/// Cache store backed by disk
+pub struct DiskCache<K, V> {
+    pub(super) seconds: u64,
+    pub(super) refresh: bool,
+    pub(super) namespace: String,
+    pub(super) prefix: String,
+    connection: sled::Db,
+    disk_path: PathBuf,
+    _phantom_k: PhantomData<K>,
+    _phantom_v: PhantomData<V>,
+}
+
+impl<K, V> DiskCache<K, V>
+where
+    K: Display,
+    V: Serialize + DeserializeOwned,
+{
+    #[allow(clippy::new_ret_no_self)]
+    /// Initialize a `DiskCacheBuilder`
+    pub fn new<S: AsRef<str>>(prefix: S, seconds: u64) -> DiskCacheBuilder<K, V> {
+        DiskCacheBuilder::new(prefix, seconds)
+    }
+
+    fn generate_key(&self, key: &K) -> String {
+        format!("{}{}{}", self.namespace, self.prefix, key)
+    }
+
+    /// Return the disk path used
+    pub fn disk_path(&self) -> PathBuf {
+        self.disk_path.clone()
+    }
+
+    fn last_cleanup(connection: &Db) -> SystemTime {
+        match connection.get(LAST_CLEANUP_KEY) {
+            Ok(Some(l)) => match rmp_serde::from_slice::<SystemTime>(&l) {
+                Ok(l) => l,
+                _ => SystemTime::UNIX_EPOCH
+            },
+            _ => SystemTime::UNIX_EPOCH
+        }
+    }
+
+    fn cleanup_expired_entries(connection: &Db) {
+        let now = SystemTime::now();
+        let last_cleanup = DiskCache::<K, V>::last_cleanup(connection);
+
+        if last_cleanup + Duration::from_secs(10) < now {
```

Comment:
Could we remove the interval logic, just have this do the full scan, and make it public? I think we might want to remove this call from the cache functions since this could end up causing every cache retrieval to do a full scan if you aren't frequently calling a function. Having it available for users to manually call would be good though so you can choose how to periodically call it (threads of async tasks)